In [1]:
# Update sklearn to prevent version mismatches
!pip install sklearn --upgrade

In [2]:
# install joblib. This will be used to save your model. 
# Restart your kernel after installing 
!pip install joblib

In [3]:
import pandas as pd

# Read the CSV and Perform Basic Data Cleaning

In [4]:
df = pd.read_csv("exoplanet_data.csv")
# Drop the null columns where all values are null
df = df.dropna(axis='columns', how='all')
# Drop the null rows
df = df.dropna()
df.head()

koi_disposition  koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
0       CONFIRMED              0              0              0              0   
1  FALSE POSITIVE              0              1              0              0   
2  FALSE POSITIVE              0              1              0              0   
3       CONFIRMED              0              0              0              0   
4       CONFIRMED              0              0              0              0   

   koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
0   54.418383     2.479000e-04    -2.479000e-04   162.513840   
1   19.899140     1.490000e-05    -1.490000e-05   175.850252   
2    1.736952     2.630000e-07    -2.630000e-07   170.307565   
3    2.525592     3.760000e-06    -3.760000e-06   171.595550   
4    4.134435     1.050000e-05    -1.050000e-05   172.979370   

   koi_time0bk_err1  ...  koi_steff_err2  koi_slogg  koi_slogg_err1  \
0          0.003520  ...             -81      4.467           0.064   
1          0.000581  ...            -176      4.544           0.044   
2          0.000115  ...            -174      4.564           0.053   
3          0.001130  ...            -211      4.438           0.070   
4          0.001900  ...            -232      4.486           0.054   

   koi_slogg_err2  koi_srad  koi_srad_err1  koi_srad_err2         ra  \
0          -0.096     0.927          0.105         -0.061  291.93423   
1          -0.176     0.868          0.233         -0.078  297.00482   
2          -0.168     0.791          0.201         -0.067  285.53461   
3          -0.210     1.046          0.334         -0.133  288.75488   
4          -0.229     0.972          0.315         -0.105  296.28613   

         dec  koi_kepmag  
0  48.141651      15.347  
1  48.134129      15.436  
2  48.285210      15.597  
3  48.226200      15.509  
4  48.224670      15.714  

[5 rows x 41 columns]

In [5]:
df.columns

Index(['koi_disposition', 'koi_fpflag_nt', 'koi_fpflag_ss', 'koi_fpflag_co',
       'koi_fpflag_ec', 'koi_period', 'koi_period_err1', 'koi_period_err2',
       'koi_time0bk', 'koi_time0bk_err1', 'koi_time0bk_err2', 'koi_impact',
       'koi_impact_err1', 'koi_impact_err2', 'koi_duration',
       'koi_duration_err1', 'koi_duration_err2', 'koi_depth', 'koi_depth_err1',
       'koi_depth_err2', 'koi_prad', 'koi_prad_err1', 'koi_prad_err2',
       'koi_teq', 'koi_insol', 'koi_insol_err1', 'koi_insol_err2',
       'koi_model_snr', 'koi_tce_plnt_num', 'koi_steff', 'koi_steff_err1',
       'koi_steff_err2', 'koi_slogg', 'koi_slogg_err1', 'koi_slogg_err2',
       'koi_srad', 'koi_srad_err1', 'koi_srad_err2', 'ra', 'dec',
       'koi_kepmag'],
      dtype='object')

In [6]:
df.describe()

koi_fpflag_nt  koi_fpflag_ss  koi_fpflag_co  koi_fpflag_ec  \
count    6991.000000    6991.000000    6991.000000    6991.000000   
mean        0.157059       0.244743       0.202975       0.125018   
std         0.363882       0.429966       0.402243       0.330763   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       1.000000   

        koi_period  koi_period_err1  koi_period_err2  koi_time0bk  \
count  6991.000000     6.991000e+03     6.991000e+03  6991.000000   
mean     56.191248     1.851122e-03    -1.851122e-03   164.488820   
std     117.570962     7.184503e-03     7.184503e-03    67.020475   
min       0.259820     1.100000e-08    -1.568000e-01   120.515914   
25%       2.620126     5.005000e-06    -2.401000e-04   132.683917   
50%       8.947426     3.300000e-05    -3.300000e-05   136.739230   
75%      34.282605     2.401000e-04    -5.005000e-06   169.937005   
max    1071.232624     1.568000e-01    -1.100000e-08  1472.522306   

       koi_time0bk_err1  koi_time0bk_err2  ...  koi_steff_err2    koi_slogg  \
count       6991.000000       6991.000000  ...     6991.000000  6991.000000   
mean           0.009340         -0.009340  ...     -161.206980     4.305049   
std            0.021989          0.021989  ...       71.448481     0.439238   
min            0.000009         -0.569000  ...    -1733.000000     0.047000   
25%            0.001145         -0.010000  ...     -197.000000     4.209000   
50%            0.003990         -0.003990  ...     -159.000000     4.436000   
75%            0.010000         -0.001145  ...     -112.000000     4.543000   
max            0.569000         -0.000009  ...        0.000000     5.364000   

       koi_slogg_err1  koi_slogg_err2     koi_srad  koi_srad_err1  \
count     6991.000000      6991.00000  6991.000000    6991.000000   
mean         0.121091        -0.14048     1.740749       0.352710   
std          0.132048         0.08199     5.903415       0.839017   
min          0.000000        -1.00700     0.109000       0.000000   
25%          0.044000        -0.19500     0.829000       0.128000   
50%          0.070000        -0.12700     0.999000       0.248000   
75%          0.149000        -0.08800     1.357000       0.357000   
max          1.472000         0.00000   180.013000      25.956000   

       koi_srad_err2           ra          dec   koi_kepmag  
count    6991.000000  6991.000000  6991.000000  6991.000000  
mean       -0.388568   292.082406    43.812143    14.271508  
std         1.907797     4.762908     3.606167     1.350802  
min      -103.825000   279.856080    36.577381     6.966000  
25%        -0.252000   288.704730    40.797760    13.455000  
50%        -0.111000   292.314760    43.679661    14.534000  
75%        -0.069000   295.888550    46.693659    15.322000  
max         0.000000   301.720760    52.336010    19.065000  

[8 rows x 40 columns]

# Select your features (columns)

In [7]:
# Set features. This will also be used as your x values.
# Planet temperature,  
selected_features = df[['koi_disposition', 'koi_fpflag_nt', 'koi_period_err2','koi_slogg_err2', 'koi_fpflag_nt']]

# Create a Train Test Split

Use `koi_disposition` for the y values

In [8]:
X = selected_features.drop("koi_disposition", axis=1)
y = selected_features["koi_disposition"]
print(X.shape, y.shape)

(6991, 4) (6991,)


For the predictors: There are 6991 rows 3 columns 

For target or label: There are 6991 rows and 1 column

#### 1st STEP: Split data into training and testing

In [9]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=66)

In [10]:
X_train.head()

koi_fpflag_nt  koi_period_err2  koi_slogg_err2  koi_fpflag_nt
4623              0    -6.491000e-04          -0.293              0
282               0    -1.240000e-05          -0.063              0
2315              1    -3.030000e-07          -0.030              1
4436              0    -1.420000e-05          -0.224              0
809               0    -1.070000e-05          -0.083              0

# Pre-processing

Scale the data using the MinMaxScaler and perform some feature selection

In [11]:
# Scale your data
from sklearn.preprocessing import MinMaxScaler
X_minmax = MinMaxScaler().fit(X_train)

X_train_minmax = X_minmax.transform(X_train)
X_test_minmax = X_minmax.transform(X_test)

# Train the Model



#### Create SVC Model

In [12]:
# Create the SVC Model
from sklearn.svm import SVC 
model = SVC(kernel='linear')
model

SVC(kernel='linear')

Fit the model to our data using the fit method

In [13]:
model.fit(X_train_minmax, y_train)
print(model)

SVC(kernel='linear')


In [14]:
print(f"Training Data Score: {model.score(X_train_minmax, y_train)}")
print(f"Testing Data Score: {model.score(X_test_minmax, y_test)}")

Training Data Score: 0.5472057982071333
Testing Data Score: 0.5543478260869565


In [15]:
print('Test Acc: %.3f' % model.score(X_test_minmax, y_test))

Test Acc: 0.554


# Hyperparameter Tuning

Use `GridSearchCV` to tune the model's parameters

In [16]:
# Train the model with GridSearch
# Create the GridSearchCV estimator along with a parameter object containing the values to adjust
from sklearn.model_selection import GridSearchCV
# C and gamma are hyperparameters that were tune
param_grid = {'C': [1, 5, 10, 50],
              'gamma': [0.0001, 0.0005, 0.001, 0.005]}
#Set up grid by calling GridSearchCV(Cross Validation) and pass name of the model, dictionary (param_grid)
#verbose = amount of output text (3 is max)
grid = GridSearchCV(model, param_grid, verbose=3)

Fit the model

In [17]:
grid.fit(X_train_minmax, y_train)

Fitting 5 folds for each of 16 candidates, totalling 80 fits
[CV 1/5] END .................C=1, gamma=0.0001;, score=0.541 total time=   0.5s
[CV 2/5] END .................C=1, gamma=0.0001;, score=0.553 total time=   0.5s
[CV 3/5] END .................C=1, gamma=0.0001;, score=0.544 total time=   0.5s
[CV 4/5] END .................C=1, gamma=0.0001;, score=0.541 total time=   0.5s
[CV 5/5] END .................C=1, gamma=0.0001;, score=0.549 total time=   0.5s
[CV 1/5] END .................C=1, gamma=0.0005;, score=0.541 total time=   0.5s
[CV 2/5] END .................C=1, gamma=0.0005;, score=0.553 total time=   0.5s
[CV 3/5] END .................C=1, gamma=0.0005;, score=0.544 total time=   0.5s
[CV 4/5] END .................C=1, gamma=0.0005;, score=0.541 total time=   0.5s
[CV 5/5] END .................C=1, gamma=0.0005;, score=0.549 total time=   0.5s
[CV 1/5] END ..................C=1, gamma=0.001;, score=0.541 total time=   0.5s
[CV 2/5] END ..................C=1, gamma=0.001;

GridSearchCV(estimator=SVC(kernel='linear'),
             param_grid={'C': [1, 5, 10, 50],
                         'gamma': [0.0001, 0.0005, 0.001, 0.005]},
             verbose=3)

In [18]:
# List the best parameters and scores for the dataset
print(grid.best_params_)
print(grid.best_score_)

{'C': 50, 'gamma': 0.0001}
0.5651358254680939


### Tune Model Parameters

Use GridSearch to tune model parameters.

In [19]:
from pprint import pprint
pprint(f"Parameters currently in use: {model.get_params()}")

("Parameters currently in use: {'C': 1.0, 'break_ties': False, 'cache_size': "
 "200, 'class_weight': None, 'coef0': 0.0, 'decision_function_shape': 'ovr', "
 "'degree': 3, 'gamma': 'scale', 'kernel': 'linear', 'max_iter': -1, "
 "'probability': False, 'random_state': None, 'shrinking': True, 'tol': 0.001, "
 "'verbose': False}")


In [20]:
# Make predictions with the hypertuned model
predictions = grid.predict(X_test_minmax)

Tune and compare at least two different classifiers.

In [21]:
# Calculate classification report
from sklearn.metrics import classification_report
print(classification_report(y_test, predictions))

                precision    recall  f1-score   support

     CANDIDATE       0.81      0.14      0.24       415
     CONFIRMED       0.44      0.48      0.46       450
FALSE POSITIVE       0.62      0.83      0.71       883

      accuracy                           0.57      1748
     macro avg       0.62      0.48      0.47      1748
  weighted avg       0.62      0.57      0.53      1748



In [22]:
print('Test Acc: %.3f' % grid.score(X_test_minmax, y_test))

Test Acc: 0.575


# Save the Model

In [23]:
# save your model by updating "your_name" with your name
# and "your_model" with your model variable
# be sure to turn this in to BCS
# if joblib fails to import, try running the command to install in terminal/git-bash
import joblib
filename = 'SVC.sav'
joblib.dump(model, filename)

['SVC.sav']